---
date: "`r Sys.Date()`"
params:
  input_samples:
    value: x
  output_rds:
    value: x
output:
  html_document:
    code_folding: hide
    theme: cerulean
    toc: yes
    number_sections: true
    toc_depth: 2
    toc_float:
      collapsed: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, fig.width = 8, fig.height = 6)
```

# Set-Up

## Libraries

```{r setup_libs}
library(Seurat)
library(data.table)
library(ggplot2)
library(knitr)
library(networkD3)
library(patchwork)
library(dplyr)
library(ggthemes)
library(rlist)
library(cowplot)
library(viridis)
library(yaml)
library(tibble)
library(future)
```

## Variables

### Update according to your dataset

```{r setup_variables, warnings=F}

# read config file in yaml format
if (!file.exists("config/config.yaml")) {
  stop("config/config.yaml does not exist. Exiting the program")
} else {
  config_file <- read_yaml("config/config.yaml")
}

for (sample_folder in strsplit(params$input_samples, " ")[[1]]) {
  if (!dir.exists(sample_folder)) {
    print(paste("WARNING: sample input folder does not exist. Your input:", sample_folder))
    stop("Quitting program.")
  }
}

# Samples to process named by their treatment conditions
samples2process <- config_file$samples2process

# Determines whether or not to regress by cell cycle in scaling step 
cell_cycle_regression <- config_file$cell_cycle_regression

# List which quantiles to map onto the summary figures
metadataQuants<- unlist(config_file$metadataQuants)

# Set filtering criterion
# Mitochondrial gene expression ratio per cell
percentMitoFilt = config_file$percentMitoFilt

# Minimum and maximum counts per cell
nCountMinFilt = config_file$nCountMinFilt
nCountMaxFilt = config_file$nCountMaxFilt

# Set how many principle components to calculate
nPCs <- config_file$nPCs

# How many principle components to use for finding multi-modal neighbors
usePCs <- config_file$usePCs

# visualize cell cycle states versus expression level of these cell cycle genes.
cc_genes <- config_file$cc_genes

# how many neighbors (k) to use when picking anchors for mapping
k.anchor <- config_file$`k.anchor`

doc_title <- paste(config_file$title, '- preprocessing')
author_list <- paste(config_file$authors, collapse = ", ")

# list of antibodies
ab_list <- config_file$antibody_list
```

---
title: "`r doc_title`"
author: "`r author_list`"
---

### Inherent Variables

```{r setup_inherent_variables}

# List to store Seurat objects for each sample
exptsList <- list()

# List to store CITE-seq antibodies sequenced in each sample
antibodies <- list()

# List to store sample metadata 
metadataList <- list()

# List of data frames to store quantile information for metadata features
metadataQuantList <- list()

# List of data frames to store number of cells within each quantile
metadataQuantCellCountsList <- list()

# Settings list to add to misc slot in Seurat objects
settingsList <- list()

# list of plots to see effects of cell cycle regression
regressionPlots <- list()

# fx to print out metadata features more efficiently in table format
summary_table <- function(samples2process, summVar, indentation, col.names) {
  for (i in c(samples2process)) {
    temp_counts <- cbind(metadataQuantList[[i]][summVar], metadataQuantCellCountsList[[i]][summVar]) %>% rownames_to_column("Percentile")
    cat(paste( paste(rep("#", indentation), collapse = "") , toupper(i), "\n") )
    print(knitr::kable(temp_counts, col.names = col.names))
    cat("\n")
  }
}

theme_set(theme_clean())
```

# Read in CellRanger data by sample

```{r cellranger_data}

for (sample in samples2process) {

  # Load in CellRanger Data
  counts_matrix = Read10X(data.dir = paste0("data/raw/", sample, "/outs/filtered_feature_bc_matrix"))
  
  # Create Seurat Object
  # Set thresholds for cell, feature inclusion
  min.cells = 3
  min.features = 200
  settingsList[["step_CreateSeuratObject"]][["min.cells"]] <- min.cells
  settingsList[["step_CreateSeuratObject"]][["min.features"]] <- min.features

  # Initialize the Seurat object with the raw (non-normalized data).
  myso <- CreateSeuratObject(counts = counts_matrix, project = sample, min.cells = min.cells, min.features = min.features)
  
  exptsList[[sample]] <- myso
}
```

# Prepare Metadata 

```{r metadata}

# Convert metadata quantile levels to percent format
percentFormat <- function(x) {
  return(paste(round(100*x, 2), "%", sep=""))
}

metadataQuantPercent <- sapply(metadataQuants, percentFormat)

for (sample in names(exptsList)) {
  myso <- exptsList[[sample]]
  
  # Quantify % mitochondrial features
  myso[["percentMito"]] <- PercentageFeatureSet(myso, pattern = "^MT-")
  
  # Add number of genes per unique molecular identifier (UMI) for each cell to metadata
  myso[['log10GenesPerUMI']] <- log10(myso[['nFeature_RNA']]) / log10(myso[['nCount_RNA']])
  
  # Extract metadata
  metadata <- myso@meta.data
  rownames(metadata) <- paste0(sample,'_',rownames(metadata))
  metadata$cells <- rownames(metadata)
  
  # Add to the metadata list
  metadataList[[sample]] <- metadata
  
  # Add to the Seurat object list
  exptsList[[sample]] <- myso
  
  # Calculate the quantiles in the metadata
  metadataNumeric <- which(sapply(metadata, is.numeric))
  metadataQuantList[[sample]] <- data.frame(sapply(metadataNumeric, function(y){
     quantile(x=unlist(metadata[,y]),
              probs = metadataQuants,
              na.rm = TRUE)}))
  
  # Calculate the number of cells in each quantile
  metadataQuantCellCounts <- data.frame(row.names = metadataQuantPercent)
  
  # Loop through the variables in the quantiles list
  for (metadataCol in colnames(metadataQuantList[[sample]])) {
    colVector <- list()
    
    # Calculate number of cells that are within the quantile
    for (level in metadataQuantPercent) {
      elem <- sum(metadata[[metadataCol]] < metadataQuantList[[sample]][level,metadataCol])
      colVector <- list.append(colVector, elem)
    }
    metadataQuantCellCounts[[metadataCol]] <- unlist(colVector)
  }
  
  # Save a column with the total number of cells
  metadataQuantCellCounts[['nCellsTotal']] <- rep(length(metadata$cells),length(metadataQuantPercent))
  
  # Add to the list with the other samples
  metadataQuantCellCountsList[[sample]] <- metadataQuantCellCounts

}

# Integrate the metadata information
integratedMetadata <- bind_rows(metadataList)
  
# Find quantiles of numeric metadata columns
# Mapped as dashed blue lines in the QC figures
integratedMetadataNumeric <- which(sapply(integratedMetadata, is.numeric))
integratedMetadataQuants <- sapply(integratedMetadataNumeric, function(y){
  quantile(x=unlist(integratedMetadata[,y]),
            metadataQuants,
            na.rm = TRUE)
})

```

# RNA-seq Summary Plots

## Number of Cells per Sample

```{r nCellsPerSample}
#Visualize the number of cell counts per sample
integratedMetadata %>%
        ggplot(aes(x=orig.ident, fill=orig.ident)) +
        labs(fill = "Sample") +
        geom_bar(alpha = 0.7) +
        xlab("Sample") +
        ylab("Number of Cells") +
        stat_count(geom = "text", colour = "grey30", size = 8, aes(label = ..count..),position=position_stack(vjust=1.10))
```

## Number of Counts per Cell

```{r nCount_RNA}
# Visualize the number UMIs/transcripts per cell
summVar <- "nCount_RNA"
summVarDesc <- "Number of Counts"

integratedMetadata %>%
        ggplot(aes(x=.data[[summVar]], fill=orig.ident, color=orig.ident)) +
        labs(fill = "Sample", color="Sample") +
        geom_density(alpha = 0.2) +
        ggtitle("Distribution of Counts Detected per Cell (Density)") +
        xlab(summVarDesc) +
        ylab("Density") +
        geom_vline(xintercept = integratedMetadataQuants[,summVar], size = 0.75, alpha = 0.5, linetype = "dashed", colour = "grey30") 

# Visualize the distribution of genes detected per cell via boxplot
integratedMetadata %>%
        ggplot(aes(x=orig.ident, y=.data[[summVar]], fill=orig.ident, color = orig.ident)) +
        labs(fill = "Sample", color = "Sample") +
        geom_boxplot(alpha = 0.7) +
        theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
        theme(plot.title = element_text(hjust=0.5, face="bold")) +
        ggtitle("Distribution of Counts Detected per Cell (Boxplot)") +
        xlab("Sample") +
        ylab(summVarDesc) +
        geom_violin(alpha = 0.1) +
        geom_hline(yintercept = integratedMetadataQuants[,summVar], size = 0.75, alpha = 0.5, linetype = "dashed", colour = "grey30")
```

### Summary table {.tabset}

```{r, results='asis'}
summary_table(samples2process, "nCount_RNA", 4, col.names = c("Percentile", "UMI Counts per Cell", "Number of Cells"))
```

## Distribution of Genes Detected per Cell

``` {r nFeature_RNA}
# Visualize the distribution of genes detected per cell via histogram
# Bimodal distributions could be indicative of low-quality cells or distinct cell populations
summVar <- "nFeature_RNA"
summVarDesc <- "Number of Unique Genes"

integratedMetadata %>%
        ggplot(aes(color=orig.ident, x=.data[[summVar]], fill= orig.ident)) +
        labs(fill = "Sample", color="Sample") +
        geom_density(alpha = 0.2) +
        ggtitle("Distribution of Genes Detected per Cell (Density)") +
        xlab(summVarDesc) +
        ylab("Density") +
        geom_vline(xintercept = integratedMetadataQuants[,"nFeature_RNA"], size = 0.75, alpha = 0.5, linetype = "dashed", colour = "grey30")

integratedMetadata %>%
        ggplot(aes(x=orig.ident, y=.data[[summVar]], fill=orig.ident, color = orig.ident)) +
        labs(fill = "Sample", color = "Sample") +
        geom_boxplot(alpha = 0.7) +
        theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
        theme(plot.title = element_text(hjust=0.5, face="bold")) +
        ggtitle("Distribution of Genes Detected per Cell (Boxplot)") +
        xlab("Sample") +
        ylab(summVarDesc) +
        geom_violin(alpha = 0.1) +
        geom_hline(yintercept = integratedMetadataQuants[,summVar], size = 0.75, alpha = 0.5, linetype = "dashed", colour = "grey30")
```

### Summary table {.tabset}

```{r, results='asis'}
summary_table(samples2process, "nFeature_RNA", 4, col.names = c("Percentile", "Number of Unique Genes", "Number of Cells"))
```

## Distribution of Mitochondrial Gene Expression per Cell

```{r percentMito}
summVar <- "percentMito"
summVarDesc <- "Mitochondrial Ratio"
integratedMetadata %>%
        ggplot(aes(color=orig.ident, x=.data[[summVar]], fill=orig.ident)) +
        labs(fill = "Sample", color="Sample") +
        geom_density(alpha = 0.2) +
        geom_vline(xintercept = integratedMetadataQuants[,summVar], size = 0.75, alpha = 0.5, linetype = "dashed", colour = "grey30") +
        ggtitle("Distribution of Mito GE per Cell (Density)") +
        xlab(summVarDesc) +
        ylab("Density")

integratedMetadata %>%
        ggplot(aes(x=orig.ident, y=.data[[summVar]], fill=orig.ident, color = orig.ident)) +
        labs(fill = "Sample", color = "Sample") +
        geom_boxplot(alpha = 0.7) +
        theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
        theme(plot.title = element_text(hjust=0.5, face="bold")) +
        ggtitle("Distribution of Mito GE per Cell (Boxplot)") +
        xlab("Sample") +
        geom_violin(alpha = 0.1) +
        geom_hline(yintercept = integratedMetadataQuants[,summVar], size = 0.75, alpha = 0.5, linetype = "dashed", colour = "grey30")
```

### Summary table {.tabset}

```{r, results='asis'}
summary_table(samples2process, "percentMito", 4, col.names = c("Percentile", "Mitochondrial Ratio", "Number of Cells"))
```

## Correlation Between Genes Detected and Number of Counts

```{r nFeature_nCount_RNA_Boxplot}
# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
integratedMetadata %>%
        ggplot(aes(x=nCount_RNA, y=nFeature_RNA, color=percentMito)) +
        geom_point() +
        scale_colour_gradient(low = "gray90", high = "black") +
        stat_smooth(method=lm) +
        scale_x_log10() +
        scale_y_log10() +
        ggtitle("Correlation Between Genes Detected and Number of Counts") +
        xlab("Number of Counts") +
        ylab("Number of Genes") +
        facet_wrap(~orig.ident)
```

## Distribution of Gene Expression Complexity

```{r geneComplexity}
# Visualize the overall complexity of the gene expression by visualizing the genes detected per UMI
summVar <- "log10GenesPerUMI"
summVarDesc <- "log10 Number of Genes per Count"
integratedMetadata %>%
        ggplot(aes(x=.data[[summVar]], color = orig.ident, fill=orig.ident)) +
        labs(fill = "Sample", color="Sample") +
        geom_density(alpha = 0.2) +
        geom_vline(xintercept = integratedMetadataQuants[,summVar], size = 0.75, alpha = 0.5, linetype = "dashed", colour = "grey30") +
        ggtitle("Gene Complexity") +
        xlab(summVarDesc) +
        ylab("Density")
```

### Summary table {.tabset}

```{r results='asis'}
summary_table(samples2process, "log10GenesPerUMI", 4, col.names = c("Percentile", "log10 Number of Genes per Count", "Number of Cells"))
```

# Filter

```{r filter}
preProcExptsList <- list()
for (sample in names(exptsList)) {
  myso <- exptsList[[sample]]
  # Filter high mitochondria cells
  settingsList[["step_filter"]][["percentMitoFilt"]] <- percentMitoFilt
  nMitoFilt <- sum(myso@meta.data$percentMito > percentMitoFilt)
  print(sprintf("%s:  Filter cells > %i%% mitochondria: %i/%i (%.1f%%)", sample, percentMitoFilt, nMitoFilt, ncol(myso), nMitoFilt/ncol(myso)*100))
  # Filter high and low RNA count cells
  settingsList[["step_filter"]][["nCountMinFilt"]] <- nCountMinFilt
  settingsList[["step_filter"]][["nCountMaxFilt"]] <- nCountMaxFilt
  nFeatFilt <- sum(myso@meta.data$nCount_RNA < nCountMinFilt) + sum(myso@meta.data$nCount_RNA > nCountMaxFilt)
  print(sprintf("%s:  Filter cells with %i > nCount > %i: %i/%i (%.1f%%)",
                sample, nCountMinFilt, nCountMaxFilt, nFeatFilt, ncol(myso), nFeatFilt/ncol(myso)*100))
  # Add settings info to the object under "misc"
  myso@misc$settings <- settingsList
  myso <- subset(myso, subset = nCount_RNA > nCountMinFilt & nCount_RNA < nCountMaxFilt & percentMito < percentMitoFilt)
  print(sprintf("Filtered %s:  %i cells, %i features remaining", sample, dim(myso)[1], dim(myso)[1]))
  preProcExptsList[[sample]] <- myso
}

rm(myso) # remove seurat object from workspace.

```

# Sample Normalization

```{r, normalize, eval = FALSE}

for (sample in samples2process) {

  myso <- preProcExptsList[[sample]]

  print(paste("Normalizing RNA for", sample))
  DefaultAssay(myso) <- 'RNA'
  myso <- SCTransform(myso) %>% FindVariableFeatures() %>% ScaleData() %>% RunPCA(npcs = nPCs)

  myso <- NormalizeData(myso, normalization.method = 'CLR', margin = 2) %>%
    ScaleData() %>%
    RunPCA(reduction.name = 'apca', nfeatures.print = length(ab_list), npcs = length(ab_list)) # apca for finding multimodal neighbors

  print(paste("Dimensions used for finding multi-modal neighbors"))

  print(paste("Finding WNN in", sample))
  myso <- FindMultiModalNeighbors(
    myso,
    reduction.list = list("pca", "apca"),
    dims.list = list(1:usePCs, 1:length(ab_list) - 1), # not sure why length(ab_list) leads to indexing error.
    modality.weight.name = "RNA.weight"
  )
  
  # save myso to list of Seurat objects
  preProcExptsList[[sample]] <- myso
}

# source: https://satijalab.org/seurat/articles/weighted_nearest_neighbor_analysis.html#wnn-analysis-of-cite-seq-rna-adt-1
```

## Elbow Plots {.tabset}

```{r elbowPlots, results='asis', eval = FALSE}

for (sample in samples2process) {

  myso = preProcExptsList[[sample]]

  elbowRNA = ElbowPlot(myso, ndims = nPCs, reduction = "pca") + labs(title = paste("Elbow plot for", sample, "Transcriptome"))

  cat(paste0("### ", toupper(sample)))

  cat("\n\n")
  print(elbowRNA)
  cat("\n\n")
}

```

```{r}
saveRDS(preProcExptsList, file = params$output_rds)
```

```{r}
sessionInfo()
```
---
date: "`r Sys.Date()`"
params:
  input_samples:
    value: x
  output_rds:
    value: x
output:
  html_document:
    code_folding: hide
    theme: cerulean
    toc: yes
    number_sections: true
    toc_depth: 2
    toc_float:
      collapsed: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, fig.width = 8, fig.height = 6)
```

# Set-Up

## Libraries

```{r setup_libs}
library(Seurat)
library(data.table)
library(ggplot2)
library(knitr)
library(networkD3)
library(patchwork)
library(dplyr)
library(ggthemes)
library(rlist)
library(cowplot)
library(viridis)
library(yaml)
library(tibble)
library(future)
```

## Variables

### Update according to your dataset

```{r setup_variables, warnings=F}

# read config file in yaml format
if (!file.exists("config/config.yaml")) {
  stop("config/config.yaml does not exist. Exiting the program")
} else {
  config_file <- read_yaml("config/config.yaml")
}

for (sample_folder in strsplit(params$input_samples, " ")[[1]]) {
  if (!dir.exists(sample_folder)) {
    print(paste("WARNING: sample input folder does not exist. Your input:", sample_folder))
    stop("Quitting program.")
  }
}

# Samples to process named by their treatment conditions
samples2process <- config_file$samples2process

# Determines whether or not to regress by cell cycle in scaling step 
cell_cycle_regression <- config_file$cell_cycle_regression

# List which quantiles to map onto the summary figures
metadataQuants<- unlist(config_file$metadataQuants)

# Set filtering criterion
# Mitochondrial gene expression ratio per cell
percentMitoFilt = config_file$percentMitoFilt

# Minimum and maximum counts per cell
nCountMinFilt = config_file$nCountMinFilt
nCountMaxFilt = config_file$nCountMaxFilt

# Set how many principle components to calculate
nPCs <- config_file$nPCs

# How many principle components to use for finding multi-modal neighbors
usePCs <- config_file$usePCs

# visualize cell cycle states versus expression level of these cell cycle genes.
cc_genes <- config_file$cc_genes

# how many neighbors (k) to use when picking anchors for mapping
k.anchor <- config_file$`k.anchor`

doc_title <- paste(config_file$title, '- preprocessing')
author_list <- paste(config_file$authors, collapse = ", ")

# list of antibodies
ab_list <- config_file$antibody_list
```

---
title: "`r doc_title`"
author: "`r author_list`"
---

### Inherent Variables

```{r setup_inherent_variables}

# List to store Seurat objects for each sample
exptsList <- list()

# List to store CITE-seq antibodies sequenced in each sample
antibodies <- list()

# List to store sample metadata 
metadataList <- list()

# List of data frames to store quantile information for metadata features
metadataQuantList <- list()

# List of data frames to store number of cells within each quantile
metadataQuantCellCountsList <- list()

# Settings list to add to misc slot in Seurat objects
settingsList <- list()

# list of plots to see effects of cell cycle regression
regressionPlots <- list()

# fx to print out metadata features more efficiently in table format
summary_table <- function(samples2process, summVar, indentation, col.names) {
  for (i in c(samples2process)) {
    temp_counts <- cbind(metadataQuantList[[i]][summVar], metadataQuantCellCountsList[[i]][summVar]) %>% rownames_to_column("Percentile")
    cat(paste( paste(rep("#", indentation), collapse = "") , toupper(i), "\n") )
    print(knitr::kable(temp_counts, col.names = col.names))
    cat("\n")
  }
}

theme_set(theme_clean())
```

# Read in CellRanger data by sample

```{r cellranger_data}

for (sample in samples2process) {

  # Load in CellRanger Data
  counts_matrix = Read10X(data.dir = paste0("data/raw/", sample, "/outs/filtered_feature_bc_matrix"))
  
  # Create Seurat Object
  # Set thresholds for cell, feature inclusion
  min.cells = 3
  min.features = 200
  settingsList[["step_CreateSeuratObject"]][["min.cells"]] <- min.cells
  settingsList[["step_CreateSeuratObject"]][["min.features"]] <- min.features

  # Initialize the Seurat object with the raw (non-normalized data).
  myso <- CreateSeuratObject(counts = counts_matrix, project = sample, min.cells = min.cells, min.features = min.features)
  
  exptsList[[sample]] <- myso
}
```

# Prepare Metadata 

```{r metadata}

# Convert metadata quantile levels to percent format
percentFormat <- function(x) {
  return(paste(round(100*x, 2), "%", sep=""))
}

metadataQuantPercent <- sapply(metadataQuants, percentFormat)

for (sample in names(exptsList)) {
  myso <- exptsList[[sample]]
  
  # Quantify % mitochondrial features
  myso[["percentMito"]] <- PercentageFeatureSet(myso, pattern = "^MT-")
  
  # Add number of genes per unique molecular identifier (UMI) for each cell to metadata
  myso[['log10GenesPerUMI']] <- log10(myso[['nFeature_RNA']]) / log10(myso[['nCount_RNA']])
  
  # Extract metadata
  metadata <- myso@meta.data
  rownames(metadata) <- paste0(sample,'_',rownames(metadata))
  metadata$cells <- rownames(metadata)
  
  # Add to the metadata list
  metadataList[[sample]] <- metadata
  
  # Add to the Seurat object list
  exptsList[[sample]] <- myso
  
  # Calculate the quantiles in the metadata
  metadataNumeric <- which(sapply(metadata, is.numeric))
  metadataQuantList[[sample]] <- data.frame(sapply(metadataNumeric, function(y){
     quantile(x=unlist(metadata[,y]),
              probs = metadataQuants,
              na.rm = TRUE)}))
  
  # Calculate the number of cells in each quantile
  metadataQuantCellCounts <- data.frame(row.names = metadataQuantPercent)
  
  # Loop through the variables in the quantiles list
  for (metadataCol in colnames(metadataQuantList[[sample]])) {
    colVector <- list()
    
    # Calculate number of cells that are within the quantile
    for (level in metadataQuantPercent) {
      elem <- sum(metadata[[metadataCol]] < metadataQuantList[[sample]][level,metadataCol])
      colVector <- list.append(colVector, elem)
    }
    metadataQuantCellCounts[[metadataCol]] <- unlist(colVector)
  }
  
  # Save a column with the total number of cells
  metadataQuantCellCounts[['nCellsTotal']] <- rep(length(metadata$cells),length(metadataQuantPercent))
  
  # Add to the list with the other samples
  metadataQuantCellCountsList[[sample]] <- metadataQuantCellCounts

}

# Integrate the metadata information
integratedMetadata <- bind_rows(metadataList)
  
# Find quantiles of numeric metadata columns
# Mapped as dashed blue lines in the QC figures
integratedMetadataNumeric <- which(sapply(integratedMetadata, is.numeric))
integratedMetadataQuants <- sapply(integratedMetadataNumeric, function(y){
  quantile(x=unlist(integratedMetadata[,y]),
            metadataQuants,
            na.rm = TRUE)
})

```

# RNA-seq Summary Plots

## Number of Cells per Sample

```{r nCellsPerSample}
#Visualize the number of cell counts per sample
integratedMetadata %>%
        ggplot(aes(x=orig.ident, fill=orig.ident)) +
        labs(fill = "Sample") +
        geom_bar(alpha = 0.7) +
        xlab("Sample") +
        ylab("Number of Cells") +
        stat_count(geom = "text", colour = "grey30", size = 8, aes(label = ..count..),position=position_stack(vjust=1.10))
```

## Number of Counts per Cell

```{r nCount_RNA}
# Visualize the number UMIs/transcripts per cell
summVar <- "nCount_RNA"
summVarDesc <- "Number of Counts"

integratedMetadata %>%
        ggplot(aes(x=.data[[summVar]], fill=orig.ident, color=orig.ident)) +
        labs(fill = "Sample", color="Sample") +
        geom_density(alpha = 0.2) +
        ggtitle("Distribution of Counts Detected per Cell (Density)") +
        xlab(summVarDesc) +
        ylab("Density") +
        geom_vline(xintercept = integratedMetadataQuants[,summVar], size = 0.75, alpha = 0.5, linetype = "dashed", colour = "grey30") 

# Visualize the distribution of genes detected per cell via boxplot
integratedMetadata %>%
        ggplot(aes(x=orig.ident, y=.data[[summVar]], fill=orig.ident, color = orig.ident)) +
        labs(fill = "Sample", color = "Sample") +
        geom_boxplot(alpha = 0.7) +
        theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
        theme(plot.title = element_text(hjust=0.5, face="bold")) +
        ggtitle("Distribution of Counts Detected per Cell (Boxplot)") +
        xlab("Sample") +
        ylab(summVarDesc) +
        geom_violin(alpha = 0.1) +
        geom_hline(yintercept = integratedMetadataQuants[,summVar], size = 0.75, alpha = 0.5, linetype = "dashed", colour = "grey30")
```

### Summary table {.tabset}

```{r, results='asis'}
summary_table(samples2process, "nCount_RNA", 4, col.names = c("Percentile", "UMI Counts per Cell", "Number of Cells"))
```

## Distribution of Genes Detected per Cell

``` {r nFeature_RNA}
# Visualize the distribution of genes detected per cell via histogram
# Bimodal distributions could be indicative of low-quality cells or distinct cell populations
summVar <- "nFeature_RNA"
summVarDesc <- "Number of Unique Genes"

integratedMetadata %>%
        ggplot(aes(color=orig.ident, x=.data[[summVar]], fill= orig.ident)) +
        labs(fill = "Sample", color="Sample") +
        geom_density(alpha = 0.2) +
        ggtitle("Distribution of Genes Detected per Cell (Density)") +
        xlab(summVarDesc) +
        ylab("Density") +
        geom_vline(xintercept = integratedMetadataQuants[,"nFeature_RNA"], size = 0.75, alpha = 0.5, linetype = "dashed", colour = "grey30")

integratedMetadata %>%
        ggplot(aes(x=orig.ident, y=.data[[summVar]], fill=orig.ident, color = orig.ident)) +
        labs(fill = "Sample", color = "Sample") +
        geom_boxplot(alpha = 0.7) +
        theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
        theme(plot.title = element_text(hjust=0.5, face="bold")) +
        ggtitle("Distribution of Genes Detected per Cell (Boxplot)") +
        xlab("Sample") +
        ylab(summVarDesc) +
        geom_violin(alpha = 0.1) +
        geom_hline(yintercept = integratedMetadataQuants[,summVar], size = 0.75, alpha = 0.5, linetype = "dashed", colour = "grey30")
```

### Summary table {.tabset}

```{r, results='asis'}
summary_table(samples2process, "nFeature_RNA", 4, col.names = c("Percentile", "Number of Unique Genes", "Number of Cells"))
```

## Distribution of Mitochondrial Gene Expression per Cell

```{r percentMito}
summVar <- "percentMito"
summVarDesc <- "Mitochondrial Ratio"
integratedMetadata %>%
        ggplot(aes(color=orig.ident, x=.data[[summVar]], fill=orig.ident)) +
        labs(fill = "Sample", color="Sample") +
        geom_density(alpha = 0.2) +
        geom_vline(xintercept = integratedMetadataQuants[,summVar], size = 0.75, alpha = 0.5, linetype = "dashed", colour = "grey30") +
        ggtitle("Distribution of Mito GE per Cell (Density)") +
        xlab(summVarDesc) +
        ylab("Density")

integratedMetadata %>%
        ggplot(aes(x=orig.ident, y=.data[[summVar]], fill=orig.ident, color = orig.ident)) +
        labs(fill = "Sample", color = "Sample") +
        geom_boxplot(alpha = 0.7) +
        theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
        theme(plot.title = element_text(hjust=0.5, face="bold")) +
        ggtitle("Distribution of Mito GE per Cell (Boxplot)") +
        xlab("Sample") +
        geom_violin(alpha = 0.1) +
        geom_hline(yintercept = integratedMetadataQuants[,summVar], size = 0.75, alpha = 0.5, linetype = "dashed", colour = "grey30")
```

### Summary table {.tabset}

```{r, results='asis'}
summary_table(samples2process, "percentMito", 4, col.names = c("Percentile", "Mitochondrial Ratio", "Number of Cells"))
```

## Correlation Between Genes Detected and Number of Counts

```{r nFeature_nCount_RNA_Boxplot}
# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
integratedMetadata %>%
        ggplot(aes(x=nCount_RNA, y=nFeature_RNA, color=percentMito)) +
        geom_point() +
        scale_colour_gradient(low = "gray90", high = "black") +
        stat_smooth(method=lm) +
        scale_x_log10() +
        scale_y_log10() +
        ggtitle("Correlation Between Genes Detected and Number of Counts") +
        xlab("Number of Counts") +
        ylab("Number of Genes") +
        facet_wrap(~orig.ident)
```

## Distribution of Gene Expression Complexity

```{r geneComplexity}
# Visualize the overall complexity of the gene expression by visualizing the genes detected per UMI
summVar <- "log10GenesPerUMI"
summVarDesc <- "log10 Number of Genes per Count"
integratedMetadata %>%
        ggplot(aes(x=.data[[summVar]], color = orig.ident, fill=orig.ident)) +
        labs(fill = "Sample", color="Sample") +
        geom_density(alpha = 0.2) +
        geom_vline(xintercept = integratedMetadataQuants[,summVar], size = 0.75, alpha = 0.5, linetype = "dashed", colour = "grey30") +
        ggtitle("Gene Complexity") +
        xlab(summVarDesc) +
        ylab("Density")
```

### Summary table {.tabset}

```{r results='asis'}
summary_table(samples2process, "log10GenesPerUMI", 4, col.names = c("Percentile", "log10 Number of Genes per Count", "Number of Cells"))
```

# Filter

```{r filter}
preProcExptsList <- list()
for (sample in names(exptsList)) {
  myso <- exptsList[[sample]]
  # Filter high mitochondria cells
  settingsList[["step_filter"]][["percentMitoFilt"]] <- percentMitoFilt
  nMitoFilt <- sum(myso@meta.data$percentMito > percentMitoFilt)
  print(sprintf("%s:  Filter cells > %i%% mitochondria: %i/%i (%.1f%%)", sample, percentMitoFilt, nMitoFilt, ncol(myso), nMitoFilt/ncol(myso)*100))
  # Filter high and low RNA count cells
  settingsList[["step_filter"]][["nCountMinFilt"]] <- nCountMinFilt
  settingsList[["step_filter"]][["nCountMaxFilt"]] <- nCountMaxFilt
  nFeatFilt <- sum(myso@meta.data$nCount_RNA < nCountMinFilt) + sum(myso@meta.data$nCount_RNA > nCountMaxFilt)
  print(sprintf("%s:  Filter cells with %i > nCount > %i: %i/%i (%.1f%%)",
                sample, nCountMinFilt, nCountMaxFilt, nFeatFilt, ncol(myso), nFeatFilt/ncol(myso)*100))
  # Add settings info to the object under "misc"
  myso@misc$settings <- settingsList
  myso <- subset(myso, subset = nCount_RNA > nCountMinFilt & nCount_RNA < nCountMaxFilt & percentMito < percentMitoFilt)
  print(sprintf("Filtered %s:  %i cells, %i features remaining", sample, dim(myso)[1], dim(myso)[1]))
  preProcExptsList[[sample]] <- myso
}

rm(myso) # remove seurat object from workspace.

```

# Sample Normalization

```{r, normalize, eval = FALSE}

for (sample in samples2process) {

  myso <- preProcExptsList[[sample]]

  print(paste("Normalizing RNA for", sample))
  DefaultAssay(myso) <- 'RNA'
  myso <- SCTransform(myso) %>% FindVariableFeatures() %>% ScaleData() %>% RunPCA(npcs = nPCs)

  myso <- NormalizeData(myso, normalization.method = 'CLR', margin = 2) %>%
    ScaleData() %>%
    RunPCA(reduction.name = 'apca', nfeatures.print = length(ab_list), npcs = length(ab_list)) # apca for finding multimodal neighbors

  print(paste("Dimensions used for finding multi-modal neighbors"))

  print(paste("Finding WNN in", sample))
  myso <- FindMultiModalNeighbors(
    myso,
    reduction.list = list("pca", "apca"),
    dims.list = list(1:usePCs, 1:length(ab_list) - 1), # not sure why length(ab_list) leads to indexing error.
    modality.weight.name = "RNA.weight"
  )
  
  # save myso to list of Seurat objects
  preProcExptsList[[sample]] <- myso
}

# source: https://satijalab.org/seurat/articles/weighted_nearest_neighbor_analysis.html#wnn-analysis-of-cite-seq-rna-adt-1
```

## Elbow Plots {.tabset}

```{r elbowPlots, results='asis', eval = FALSE}

for (sample in samples2process) {

  myso = preProcExptsList[[sample]]

  elbowRNA = ElbowPlot(myso, ndims = nPCs, reduction = "pca") + labs(title = paste("Elbow plot for", sample, "Transcriptome"))

  cat(paste0("### ", toupper(sample)))

  cat("\n\n")
  print(elbowRNA)
  cat("\n\n")
}

```

```{r}
saveRDS(preProcExptsList, file = params$output_rds)
```

```{r}
sessionInfo()
```
